### DISCUSSION TITLE
OpenAmp with C++

### DISCUSSION THREAD

**Discussion by shivi-mangal** [OpenAMP/libmetal#144] [General]
identifier "memory_order" is undefined 
Path: Third_Party\OpenAMP\libmetal\lib\include\metal\io.h  
line number 42

**Comment by arnopo** [comment]
Hi @shivi-mangal 
I transferred the discussion in Libmetal.

Could you specify your environment?
- target
- compiler
- libmetal version used

Then could you check if  you have following patch  integrated in last release?
https://github.com/OpenAMP/libmetal/commit/8dfbbadc2d92fb06ba0ee0bf1837b281e7a021c8

**Reply by shivi-mangal** [reply]
Target board: STM32MP1
Compiler : EWARM 
Lib metal version: 2020

**Comment by shivi-mangal** [comment]
Thanks @arnopo,
i do not have the patch integrated in the latest release.
It would be really helpful if you could provide me the link to the latest release?

there are two atomic.h files.
1) ThirdParty\OpenAMP\libmetal\lib\include\metal\atomic.h
2)ThirdParty\OpenAMP\libmetal\lib

the proposed changes in the commit are to be made in which file?

**Reply by arnopo** [reply]
> Thanks @arnopo,
> i do not have the patch integrated in the latest release.

if you are using the ST distribution, it is based on the libmetal official release v2018.10 that not properly support C++.

> It would be really helpful if you could provide me the link to the latest release?

The latest release should be available on ST GitHub in Week 13.

If you cannot/don't want to wait, I  attached 2 patches that you could apply on top of the libmetal master branch of the official project (so not on top of the the ST distrib) and then replace the ST libmetal by it. But I do not tested it so no guaranty that this works...

[patch for ST distribution.zip](https://github.com/OpenAMP/libmetal/files/5813957/patch.for.ST.distribution.zip)

> 
> there are two atomic.h files.
> 
> 1. ThirdParty\OpenAMP\libmetal\lib\include\metal\atomic.h
>    2)ThirdParty\OpenAMP\libmetal\lib
> 
> the proposed changes in the commit are to be made in which file?

In ST distribution the libmetal code has been pre-configured because some IDE does not support cmake.
The path: Third_Party\OpenAMP\libmetal\lib\include\ is the result of the cmake, which does a copy. So to be sure you can change both.

Hope that this will help you

**Reply by shivi-mangal** [reply]
Hi @arnopo 
Integrating the patch gives me Extern "C" linkage errors.
The compilation is still an issue.
Could you please help me with a work around here?

thanks.

**Reply by arnopo** [reply]
Did you replaced first the ST libmetal by the https://github.com/OpenAMP/libmetal.git repo and then apply patches i provided on top of? 

Then traces will be welcome to better understand the issue. 

But a the end , as this dicussion seems related to C++ compilation of the ST distribution with IAR, i would suggest you to create a post  in https://community.st.com/ if you need more support. I think you are the first to try this.
